ucp/_libs/ucx_endpoint.pyx

```diff
@@ -54,6 +54,13 @@ cdef int _cancel_am_recv(UCXWorker worker, uintptr_t handle_as_int):
                 **cb_kwargs
             )
 
+        # Prevent endpoint canceling AM messages multiple times.  This is important
+        # because UCX may reuse the same endpoint handle, and if a message is canceled
+        # during then endpoint finalizer, a message received on the same (new) endpoint
+        # handle may be canceled incorrectly.
+        del worker._am_recv_wait[handle_as_int]
+        worker._inflight_msgs_to_cancel["am"].discard(handle_as_int)
```

Comment:
```suggestion
        # Prevent endpoint canceling AM messages multiple times. This is important
        # because UCX may reuse the same endpoint handle, and if a message is canceled
        # during then endpoint finalizer, a message received on the same (new) endpoint
        # handle may be canceled incorrectly.
        del worker._am_recv_wait[handle_as_int]
        worker._inflight_msgs_to_cancel["am"].discard(handle_as_int)
```

Reply:
Ha! Good catch!